### PR TITLE
GG-35559 Save cache store metadata after recovery

### PR DESCRIPTION
**Pull request by sashapolo**
https://ggsystems.atlassian.net/browse/GG-35559

### CODE REVIEW

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDynamicSqlRestoreTest.java [hunk 1]
```diff
@@ -273,6 +275,103 @@ public void testIndexCreationWhenNodeStopped() throws Exception {
...
+     * Tests a scenario when an index gets created when a node is turned off and no checkpoint is issued upon the node
+     * stop.
+     */
+    @WithSystemProperty(key = IGNITE_PDS_SKIP_CHECKPOINT_ON_NODE_STOP, value = "true")
+    @Test
+    public void testIndexCreationWhenNodeIsStoppedWithoutCheckpoint() throws Exception {
+        // Start topology.
+        Ignite ignite = startGrids(2);
+
+        ignite.cluster().active(true);
```
Comment (ibessonov) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDynamicSqlRestoreTest.java:288]: I thought this API is deprecated

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDynamicSqlRestoreTest.java [hunk 2]
```diff
@@ -273,6 +275,103 @@ public void testIndexCreationWhenNodeStopped() throws Exception {
...
+
+        // Create table, add some data.
+        int entryCnt = 50;
+
+        try (Connection conn = DriverManager.getConnection("jdbc:ignite:thin://127.0.0.1:10800")) {
```
Comment (ibessonov) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDynamicSqlRestoreTest.java:293]: What's the reason of using a JDBC here? Could this be replicated with "native" SQL API?

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDynamicSqlRestoreTest.java [hunk 3]
```diff
@@ -273,6 +275,103 @@ public void testIndexCreationWhenNodeStopped() throws Exception {
...
+        // Create an index on remaining node.
+        try (Connection conn = DriverManager.getConnection("jdbc:ignite:thin://127.0.0.1:10800")) {
+            executeJdbc(conn, "create index PERSON_FIRST_NAME_IDX on PERSON(FIRST_NAME)");
+        }
+
+        // Restart second node.
+        startGrid(1);
+
+        // Await for index rebuild on started node.
+        assert GridTestUtils.waitForCondition(() -> {
```
Comment (ibessonov) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDynamicSqlRestoreTest.java:333]: Please use assertTrue, it's a bad habit to have a code with side effects in "assert" statement.

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDynamicSqlRestoreTest.java [hunk 4]
```diff
@@ -273,6 +275,103 @@ public void testIndexCreationWhenNodeStopped() throws Exception {
...
+                    stmt.setString(1, String.valueOf(1));
+
+                    StringBuilder fullPlan = new StringBuilder();
+
+                    try (ResultSet rs = stmt.executeQuery()) {
+                        while (rs.next())
+                            fullPlan.append(rs.getString(1)).append("; ");
+                    }
+
+                    System.out.println("PLAN: " + fullPlan);
```
Comment (ibessonov) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDynamicSqlRestoreTest.java:346]: Should this be left in the code?

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDynamicSqlRestoreTest.java [hunk 5]
```diff
@@ -273,6 +275,103 @@ public void testIndexCreationWhenNodeStopped() throws Exception {
...
+
+                    return fullPlan.toString().contains("PUBLIC.PERSON_FIRST_NAME_IDX");
+                }
+            }
+            catch (Exception e) {
+                throw new RuntimeException("Query failed.", e);
```
Comment (ibessonov) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDynamicSqlRestoreTest.java:352]: I'd rather have a `fail(...)` here

Comment (sashapolo) [reply modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteDynamicSqlRestoreTest.java:352]: we will lose the original exception then

**Review by ibessonov — APPROVED** [review]

**Pull request merged**